src/main/java/org/openjdk/jextract/impl/ToplevelBuilder.java

```diff
@@ -77,23 +77,41 @@ private static HeaderFileBuilder createFirstHeader(SourceFileBuilder sfb, List<S
         } else {
             first.appendIndentedLines("public static final ValueLayout.OfLong C_LONG = ValueLayout.JAVA_LONG;");
         }
+        first.emitRuntimeHelperMethods();
         return first;
     }
 
     public List<JavaFileObject> toFiles() {
-        boolean hasOneHeader = lastHeader == firstHeader;
-        if (hasOneHeader) {
-            firstHeader.emitRuntimeHelperMethods();
-        }
         lastHeader.classEnd();
 
         List<JavaFileObject> files = new ArrayList<>();
-        files.add(firstHeader.sourceFileBuilder().toFile(s -> s.replace("extends #{SUPER}",
-                hasOneHeader ? "" : "extends " + lastHeader.className())));
-        files.addAll(builders.stream()
+
+        if (headerBuilders.size() == 1) {
```

Comment:
The changes here are the trickiest. Basically, we can't figure out in advance which name header classes should have. So we end up generating something like:

```
<headerName>_h#{SUFFIX} extends {PREV_SUFFIX}
```

And then leave it to the code here to fixup all the suffixes. For this reason, the name of the runtime helper class cannot just be the current "class name" as that would include the `#{SUFFIX}` bit - but we know that there is going to be _some_ `<headerName>_h` generated at some point, so we can just use that as a runtime helper name.

Overall, I'm not too proud of this code, but at the same time, after evaluating a couple of alternatives they all seemed too complicated. And, we were doing some magic name adaptation even before.

Reply:
> but we know that there is going to be some <headerName>_h generated at some point, so we can just use that as a runtime helper name.

So, just to clarify: we use the last header class we generate as the runtime helper (having no name suffix), but the runtime helper methods and primitive layouts are actually in the first header, and they are visible through static inheritance?

Reply:
yes, the last (deepest in the inheritance tree) is used as the runtime helper. That is also the public header class that users depend on. The helper methods are not here, but they are accessible anyway because of inheritance, so it works out.